app/workers/email_archive_worker.rb

```diff
@@ -0,0 +1,71 @@
+class EmailArchiveWorker
+  include Sidekiq::Worker
+
+  LOCK_NAME = "email_archive_worker".freeze
+
+  def perform
+    Email.with_advisory_lock(LOCK_NAME, timeout_seconds: 0) do
+      EmailArchiveQuery.call.in_batches do |batch|
+        Email.transaction { archive_batch(batch) }
+      end
+    end
+  end
+
+private
+
+  def archive_batch(batch)
+    archived_at = Time.zone.now
+
+    import_batch(batch, archived_at)
+    Email.where(id: batch.pluck(:id)).update_all(archived_at: archived_at)
+  end
+
+  def import_batch(batch, archived_at)
+    to_import = batch.as_json.map { |e| build_email_archive(e, archived_at) }
+    columns = to_import.first.keys.map(&:to_s)
+    values = to_import.map(&:values)
+    EmailArchive.import!(columns, values)
+  end
+
+  def build_email_archive(email_data, archived_at)
+    content_change = build_content_change(email_data)
+
+    {
+      archived_at: archived_at,
+      content_change: content_change,
+      created_at: email_data["created_at"],
+      finished_sending_at: email_data["finished_sending_at"],
+      id: email_data["id"],
+      sent: email_data["sent"],
+      subject: email_data["subject"],
+      subscriber_id: build_subscriber_id(email_data),
+    }
+  end
+
+  def build_content_change(email_data)
+    return if email_data["content_change_ids"].empty?
+
+    if email_data["digest_run_ids"].count > 1
```

Comment:
What circumstances do you think would cause us to have email associated with more than one digest run?

Reply:
Should be never. At an application level it should be impossible, but the means that we use to pull the data from the db means that it is an array that db wise could be multiple values. It's the same situation with Subscribers.

I was a bit unsure what to do with that possibility so @thomasleese and I had a chat through it and thought just use first value but moan to raven if we get anything weird (as that's probably a clue something weird is going on at an app level)

Reply:
Cool thanks. I expect you've already wrestled with this, but the archive process feels like a bit on an anomalous place to be integrity checking our data. There's nowhere else to put it at the moment though.